VMware: new module vmware_dvswitch_nioc

vmwjoseph opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Adding new modules to configure Network IO Control on DVS
- `vmware_dvswitch_nioc` is used to enable, disable, and control the version of Network IO Control


<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, task or feature -->
- vmware_dvswitch_nioc

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.6.1
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/opt/lib/ansible/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.14 (default, Sep 23 2017, 22:06:14) [GCC 7.2.0]
```


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
https://gitlab.eng.vmware.com/cpe-cloudops/Nova/ansible-modules/merge_requests/16
https://gitlab.eng.vmware.com/cpe-cloudops/Nova/vsphere/merge_requests/81
https://gitlab.eng.vmware.com/cpe-cloudops/Nova/core/merge_requests/32

# Enable NIOC
```
changed: [10.161.56.129] => {
    "changed": true, 
    "invocation": {
        "module_args": {
            "hostname": "10.161.56.129", 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "port": 443, 
            "resources": [
                {
                    "limit": -1, 
                    "name": "management", 
                    "reservation": 0, 
                    "shares": 10, 
                    "shares_level": "normal"
                }, 
                {
                    "limit": -1, 
                    "name": "virtualMachine", 
                    "reservation": 0, 
                    "shares": 60, 
                    "shares_level": "custom"
                }, 
                {
                    "limit": -1, 
                    "name": "vmotion", 
                    "reservation": 0, 
                    "shares": 50, 
                    "shares_level": "normal"
                }, 
                {
                    "limit": -1, 
                    "name": "vsan", 
                    "reservation": 0, 
                    "shares": 50, 
                    "shares_level": "high"
                }
            ], 
            "state": "present", 
            "switch": "nova_dvs", 
            "username": "administrator@vsphere.local", 
            "validate_certs": false, 
            "version": "version2"
        }
    }, 
    "resources_changed": [
        "management", 
        "virtualMachine"
    ], 
    "result": "Enabled NIOC with version version2"
}
```


# Change version2 to version3
```
changed: [10.161.56.129] => {
    "changed": true, 
    "invocation": {
        "module_args": {
            "hostname": "10.161.56.129", 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "port": 443, 
            "resources": [
                {
                    "limit": -1, 
                    "name": "management", 
                    "reservation": 0, 
                    "shares": 10, 
                    "shares_level": "normal"
                }, 
                {
                    "limit": -1, 
                    "name": "virtualMachine", 
                    "reservation": 0, 
                    "shares": 60, 
                    "shares_level": "custom"
                }, 
                {
                    "limit": -1, 
                    "name": "vmotion", 
                    "reservation": 0, 
                    "shares": 50, 
                    "shares_level": "normal"
                }, 
                {
                    "limit": -1, 
                    "name": "vsan", 
                    "reservation": 128, 
                    "shares": 50, 
                    "shares_level": "high"
                }
            ], 
            "state": "present", 
            "switch": "nova_dvs", 
            "username": "administrator@vsphere.local", 
            "validate_certs": false, 
            "version": "version3"
        }
    }, 
    "resources_changed": [
        "vsan"
    ], 
    "result": "Set NIOC to version version3"
}
```

# Change resources
```
changed: [10.161.56.129] => {
    "changed": true, 
    "invocation": {
        "module_args": {
            "hostname": "10.161.56.129", 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "port": 443, 
            "resources": [
                {
                    "limit": -1, 
                    "name": "management", 
                    "reservation": 0, 
                    "shares": 30, 
                    "shares_level": "custom"
                }, 
                {
                    "limit": -1, 
                    "name": "virtualMachine", 
                    "reservation": 0, 
                    "shares": 60, 
                    "shares_level": "high"
                }, 
                {
                    "limit": -1, 
                    "name": "vmotion", 
                    "reservation": 0, 
                    "shares": 50, 
                    "shares_level": "normal"
                }, 
                {
                    "limit": -1, 
                    "name": "vsan", 
                    "reservation": 100, 
                    "shares": 50, 
                    "shares_level": "high"
                }
            ], 
            "state": "present", 
            "switch": "nova_dvs", 
            "username": "administrator@vsphere.local", 
            "validate_certs": false, 
            "version": "version3"
        }
    }, 
    "resources_changed": [
        "management", 
        "virtualMachine", 
        "vsan"
    ], 
    "result": "Resource configuration modified"
}
```

# Rerun previous (should be no chnages)
```
ok: [10.161.56.129] => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "hostname": "10.161.56.129", 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "port": 443, 
            "resources": [
                {
                    "limit": -1, 
                    "name": "management", 
                    "reservation": 0, 
                    "shares": 30, 
                    "shares_level": "custom"
                }, 
                {
                    "limit": -1, 
                    "name": "virtualMachine", 
                    "reservation": 0, 
                    "shares": 60, 
                    "shares_level": "high"
                }, 
                {
                    "limit": -1, 
                    "name": "vmotion", 
                    "reservation": 0, 
                    "shares": 50, 
                    "shares_level": "normal"
                }, 
                {
                    "limit": -1, 
                    "name": "vsan", 
                    "reservation": 100, 
                    "shares": 50, 
                    "shares_level": "high"
                }
            ], 
            "state": "present", 
            "switch": "nova_dvs", 
            "username": "administrator@vsphere.local", 
            "validate_certs": false, 
            "version": "version3"
        }
    }, 
    "resources_changed": [], 
    "result": "Unchanged"
}
```

# Change version3 to version2
```
changed: [10.161.56.129] => {
    "changed": true, 
    "invocation": {
        "module_args": {
            "hostname": "10.161.56.129", 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "port": 443, 
            "resources": [
                {
                    "limit": -1, 
                    "name": "management", 
                    "reservation": 0, 
                    "shares": 30, 
                    "shares_level": "custom"
                }, 
                {
                    "limit": -1, 
                    "name": "virtualMachine", 
                    "reservation": 0, 
                    "shares": 60, 
                    "shares_level": "high"
                }, 
                {
                    "limit": -1, 
                    "name": "vmotion", 
                    "reservation": 0, 
                    "shares": 50, 
                    "shares_level": "normal"
                }, 
                {
                    "limit": -1, 
                    "name": "vsan", 
                    "reservation": 100, 
                    "shares": 50, 
                    "shares_level": "high"
                }
            ], 
            "state": "present", 
            "switch": "nova_dvs", 
            "username": "administrator@vsphere.local", 
            "validate_certs": false, 
            "version": "version2"
        }
    }, 
    "resources_changed": [
        "management", 
        "vsan"
    ], 
    "result": "Set NIOC to version version2"
}
```

# Disable NIOC
```
changed: [10.161.56.129] => {
    "changed": true, 
    "invocation": {
        "module_args": {
            "hostname": "10.161.56.129", 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "port": 443, 
            "resources": [
                {
                    "limit": -1, 
                    "name": "management", 
                    "reservation": 100, 
                    "shares": 10, 
                    "shares_level": "high"
                }, 
                {
                    "limit": -1, 
                    "name": "virtualMachine", 
                    "reservation": 0, 
                    "shares": 30, 
                    "shares_level": "custom"
                }, 
                {
                    "limit": -1, 
                    "name": "vmotion", 
                    "reservation": 0, 
                    "shares": 50, 
                    "shares_level": "normal"
                }, 
                {
                    "limit": -1, 
                    "name": "vsan", 
                    "reservation": 0, 
                    "shares": 50, 
                    "shares_level": "high"
                }
            ], 
            "state": "absent", 
            "switch": "nova_dvs", 
            "username": "administrator@vsphere.local", 
            "validate_certs": false, 
            "version": "version2"
        }
    }, 
    "resources_changed": [], 
    "result": "Disabled NIOC"
}
```

Commit:
VMware: Add new module vmware_dvswitch_nioc

Signed-off-by: Joseph Andreatta <joseph@vmware.com>
Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,189 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright 2018 VMware, Inc.
```
```# Copyright: (c) 2018, VMware, Inc.```

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,189 @@
+module: vmware_dvswitch_nioc
+short_description: Manage distributed vSwitch Network IO Control
+description:
+    - Manages distributed vSwitch Network IO Control
```
period at the end of statement.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,189 @@
+options:
+    switch_name:
+        description:
+            - The name of the switch to create or remove
```
period at the end of statement.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,189 @@
+EXAMPLES = '''
+- name: Enable NIOC
+  local_action:
+    module: vmware_dvswitch_nioc
```
remove `local_action` and call directly `module`

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,189 @@
+- name: Enable NIOC
+  local_action:
+    module: vmware_dvswitch_nioc
+    hostname: vcenter_ip_or_hostname
```
```
    hostname: '{{ vcenter_hostname }}'
    username: '{{ vcenter_username }}'
    password: '{{ vcenter_password }}'
```

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,189 @@
+    password: vcenter_password
+    switch_name: dvSwitch
+    version: version3
+    state: present
```
Add example of disable as well.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,189 @@
+    HAS_PYVMOMI = False
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.vmware import (HAS_PYVMOMI,
```
Could you please use this https://github.com/ansible/ansible/blob/f9d4e4d3961d36f127c290ea23d603b03b079ddf/lib/ansible/module_utils/vmware.py#L770
instead of `object` class ?

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,189 @@
+
+    def set_nioc_version(self):
+        valid_versions = ['version2', 'version3']
+        if self.version not in valid_versions:
```
This is handled by `argument_spec`

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,189 @@
+    except vmodl.MethodFault as method_fault:
+        module.fail_json(msg=method_fault.msg)
+    except Exception as e:
+        module.fail_json(msg=str(e))
```
use `to_native` instead of `str`

```python
from ansible.module_utils._text import to_native
...
msg=to_native(e))
```

Akasurde reviewed (changes requested):
Could you please check if this works with vcsim and add testcase for this change ? You can check the existing testcases [here](https://github.com/ansible/ansible/tree/devel/test/integration/targets/).

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,189 @@
+        result = "Enable NIOC"
+
+        if not self.module.check_mode:
+            task = self.dvs.EnableNetworkResourceManagement(enable=True)
```
If you provide non-existing DVSwitch then `self.dvs` will be `None` and module will fail with error saying 

```
"'NoneType' object has no attribute 'EnableNetworkResourceManagement'"
```

pdellaert replied:
This should be monitoring for the task to be successful, and if not, exit with a failure and error message (wait_for_task with try/catch and exit in failure)

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,189 @@
+        upgrade_spec.networkResourceControlVersion = self.version
+        task = self.dvs.ReconfigureDvs_Task(spec=upgrade_spec)
+
+        changed, result = wait_for_task(task)
```
might want to wrap `wait_for_task` with try except block to handle exception.

pdellaert replied:
And exit with a proper failure message if it fails.

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,189 @@
+        required: False
+    state:
+        description:
+            - Create or remove dvSwitch
```
`Create or remove a dvSwitch.`

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,189 @@
+        result = "Disable NIOC"
+
+        if not self.module.check_mode:
+            task = self.dvs.EnableNetworkResourceManagement(enable=False)
```
This should be monitoring for the task to be successful, and if not, exit with a failure and error message (wait_for_task with try/catch and exit in failure)

vmwjoseph replied:
Unfortunately `EnableNetworkResourceManagement` doesn't return a task. But I can add the error handling around the call.

https://code.vmware.com/apis/358/vsphere#/doc/vim.DistributedVirtualSwitch.html#enableNetworkResourceManagement

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,189 @@
+        self.dvs = find_dvs_by_name(self.content, self.switch_name)
+
+        if self.dvs is None:
+            return 'absent'
```
This should exit in failure

pdellaert reviewed (changes requested):
Few items top of @Akasurde's comments, overall it's a great start. 

Even if vcsim doesn't support this, i'd suggest making the test cases and commenting them (look at some other test cases for unsupported features in vcsim).

Thanks!

vmwjoseph commented:
@Akasurde @pdellaert Thank you for reviewing! I believe I have addressed all of your feedback. I've updated the `ADDITIONAL INFORMATION` rerunning the test cases.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,205 @@
+short_description: Manage distributed vSwitch Network IO Control.
+description:
+    - Manages distributed vSwitch Network IO Control.
+version_added: 2.7
```
We are beyond 2.7 development cycle, we need to update this to 2.8 as it is current development cycle.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,205 @@
+
+    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=True)
+
+    if not HAS_PYVMOMI:
```
This is handled in `PyVmomi` helper class.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,205 @@
+class VMwareDVSwitchNIOC(PyVmomi):
+
+    def __init__(self, module):
+        self.module = module
```
```
        super(VMwareDVSwitchNIOC, self).__init__(module)
```
is missing.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,205 @@
+        self.switch_name = self.module.params['switch_name']
+        self.version = self.module.params['version']
+        self.state = self.module.params['state']
+        self.content = connect_to_api(module)
```
This is not required if using `super`

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,201 @@
+module: vmware_dvswitch_nioc
+short_description: Manage distributed vSwitch Network IO Control.
+description:
+    - Manages distributed vSwitch Network IO Control.
```
```
This module can be used to manage distributed vSwitch Network IO control configurations.

vmwjoseph replied:
Thanks, I've changed the description.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,201 @@
+options:
+    switch_name:
+        description:
+            - The name of the switch to create or remove.
```
We are not dealing with Standard switch, right ? Also, we are not creating or removing switch.

vmwjoseph replied:
Correct, this functionality is only for distributed switches. I've fixed the description.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,201 @@
+        required: False
+    state:
+        description:
+            - Create or remove a dvSwitch.
```
Change description to match `enable` and `disable` dvSwitch.

vmwjoseph replied:
Thanks for catching this.

ckotte commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,201 @@
+extends_documentation_fragment: vmware.documentation
+'''
+
+RETURN = ''' # '''
```
Could you add what will be returned?

vmwjoseph replied:
There really isn't much returned, but I've added this.

ckotte commented on lib/ansible/modules/cloud/vmware/vmware_dvswitch_nioc.py:
```diff
@@ -0,0 +1,201 @@
+    - "python >= 2.6"
+    - PyVmomi
+options:
+    switch_name:
```
Could you please change this to `switch`. We don't want to use `_name` anymore.

vmwjoseph replied:
Sure, I've changed this.

ckotte commented:
LGTM

@vmwjoseph Any plans to add resource allocation / network resource pool configuration to this module?

vmwjoseph commented:
> LGTM
> 
> @vmwjoseph Any plans to add resource allocation / network resource pool configuration to this module?

Yes, I have just submitted https://github.com/ansible/ansible/pull/48688 for the resources

Akasurde commented:
@pdellaert Could you please check again so that we can proceed with this PR ? Thanks.

pdellaert reviewed (approved):
LGTM

Akasurde commented:
@vmwjoseph Thanks for the contribution.
